extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafkainput/KafkaInputReader.java

```diff
@@ -0,0 +1,196 @@
+      String keyColumnPrefix,
+      String recordTimestampColumnPrefix
+  )
+  {
+    this.inputRowSchema = inputRowSchema;
+    this.record = record;
+    this.headerParser = headerParser;
+    this.keyParser = keyParser;
+    this.valueParser = valueParser;
+    this.keyColumnPrefix = keyColumnPrefix;
+    this.recordTimestampColumnPrefix = recordTimestampColumnPrefix;
+  }
+
+  private CloseableIterator<InputRow> buildBlendedRows(InputEntityReader valueParser, Map<String, Object> headerKeyList) throws IOException
+  {
+    return valueParser.read().map(
+        r -> {
+          MapBasedInputRow valueRow;
+          try {
+            // Return type for the value parser should be of type MapBasedInputRow
+            // Parsers returning other types are not compatible currently.
+            valueRow = (MapBasedInputRow) r;
+          }
+          catch (ClassCastException e) {
+            throw new ParseException("Unsupported input format in valueFormat. KafkaInputformat only supports input format that return MapBasedInputRow rows");
+          }
+          Map<String, Object> event = new HashMap<>(headerKeyList);
+          /* Currently we prefer payload attributes if there is a collision in names.
+              We can change this beahvior in later changes with a config knob. This default
+              behavior lets easy porting of existing inputFormats to the new one without any changes.
+            */
+          event.putAll(valueRow.getEvent());
+
+          HashSet<String> newDimensions = new HashSet<String>(valueRow.getDimensions());
+          newDimensions.addAll(headerKeyList.keySet());
+          // Remove the dummy timestamp added in KafkaInputFormat
+          newDimensions.remove("__kif_auto_timestamp");
+
+          final List<String> schemaDimensions = inputRowSchema.getDimensionsSpec().getDimensionNames();
+          final List<String> dimensions;
+          if (!schemaDimensions.isEmpty()) {
+            dimensions = schemaDimensions;
+          } else {
+            dimensions = Lists.newArrayList(
+                Sets.difference(newDimensions, inputRowSchema.getDimensionsSpec().getDimensionExclusions())
+            );
+          }
+
+          return new MapBasedInputRow(
+              inputRowSchema.getTimestampSpec().extractTimestamp(event),
+              dimensions,
+              event
+          );
+        }
+    );
+  }
+
+  private CloseableIterator<InputRow> buildRowsWithoutValuePayload(Map<String, Object> headerKeyList)
+  {
+    HashSet<String> newDimensions = new HashSet<String>(headerKeyList.keySet());
+    final List<String> schemaDimensions = inputRowSchema.getDimensionsSpec().getDimensionNames();
+    final List<String> dimensions;
+    if (!schemaDimensions.isEmpty()) {
+      dimensions = schemaDimensions;
+    } else {
+      dimensions = Lists.newArrayList(
+          Sets.difference(newDimensions, inputRowSchema.getDimensionsSpec().getDimensionExclusions())
+      );
+    }
```

Comment:
we seem to be repeating this code twice, any chance we can avoid this repetition?

Reply:
done